core/src/main/java/org/apache/iceberg/avro/InternalReader.java

```diff
@@ -0,0 +1,251 @@
+                new ResolvingReadBuilder(),
+                AccessByID.instance());
+  }
+
+  @Override
+  public void setSchema(Schema schema) {
+    this.fileSchema = schema;
+    initReader();
+  }
+
+  public InternalReader<T> setRootType(Class<? extends StructLike> rootClass) {
+    typeMap.put(-1, rootClass);
+    return this;
+  }
+
+  public InternalReader<T> setCustomType(int fieldId, Class<? extends StructLike> structClass) {
+    typeMap.put(fieldId, structClass);
+    return this;
+  }
+
+  @Override
+  public void setRowPositionSupplier(Supplier<Long> posSupplier) {
+    if (reader instanceof SupportsRowPosition) {
+      ((SupportsRowPosition) reader).setRowPositionSupplier(posSupplier);
+    }
+  }
+
+  @Override
+  public T read(T reuse, Decoder decoder) throws IOException {
+    return reader.read(decoder, reuse);
+  }
+
+  private class ResolvingReadBuilder
+      extends AvroWithPartnerVisitor<Pair<Integer, Type>, ValueReader<?>> {
+    @Override
+    public ValueReader<?> record(
+        Pair<Integer, Type> partner, Schema record, List<ValueReader<?>> fieldResults) {
+      if (partner == null) {
+        return ValueReaders.skipStruct(fieldResults);
+      }
+
+      Types.StructType expected = partner.second().asStructType();
+      List<Pair<Integer, ValueReader<?>>> readPlan =
+          ValueReaders.buildReadPlan(expected, record, fieldResults, idToConstant);
+
+      return structReader(readPlan, partner.first(), expected);
+    }
+
+    private ValueReader<?> structReader(
+        List<Pair<Integer, ValueReader<?>>> readPlan, int fieldId, Types.StructType struct) {
+
+      Class<? extends StructLike> structClass = typeMap.get(fieldId);
+      if (structClass != null) {
+        return InternalReaders.struct(struct, structClass, readPlan);
+      } else {
+        return InternalReaders.struct(struct, readPlan);
+      }
+    }
+
+    @Override
+    public ValueReader<?> union(
+        Pair<Integer, Type> partner, Schema union, List<ValueReader<?>> options) {
+      return ValueReaders.union(options);
+    }
+
+    @Override
+    public ValueReader<?> arrayMap(
+        Pair<Integer, Type> partner,
+        Schema map,
+        ValueReader<?> keyReader,
+        ValueReader<?> valueReader) {
+      return ValueReaders.arrayMap(keyReader, valueReader);
+    }
+
+    @Override
+    public ValueReader<?> array(
+        Pair<Integer, Type> partner, Schema array, ValueReader<?> elementReader) {
+      return ValueReaders.array(elementReader);
+    }
+
+    @Override
+    public ValueReader<?> map(Pair<Integer, Type> partner, Schema map, ValueReader<?> valueReader) {
+      return ValueReaders.map(ValueReaders.strings(), valueReader);
+    }
+
+    @Override
+    public ValueReader<?> primitive(Pair<Integer, Type> partner, Schema primitive) {
+      LogicalType logicalType = primitive.getLogicalType();
+      if (logicalType != null) {
+        switch (logicalType.getName()) {
+          case "date":
```

Comment:
These string constants seem fragile but I don't have a better suggestion.

Reply:
I think it's cleaner this way. No need to introduce constants that are not widely used and would need to be public for any reuse.